### PR TITLE
TF-2085 Combine conditions to create rules

### PR DESCRIPTION
**Pull request by hieutbui**
## Issue
---
 #2085

## Resolved
---

https://github.com/linagora/tmail-flutter/assets/80142234/901fea09-855d-4af9-a06f-0ed6e6a81d19

https://github.com/linagora/tmail-flutter/assets/80142234/a925cd82-e8dd-4f1a-afee-8129ffc5063d

### CODE REVIEW

File: lib/features/rules_filter_creator/presentation/widgets/rule_condition_combiner_bottomsheet_action_tile_builder.dart [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            ? MaterialStateMouseCursor.clickable
+            : MouseCursor.defer,
+        child: CupertinoActionSheetAction(
+          key: key,
+          child: Stack(
+            children: [
+              Center(
+                child: Text(actionName,
+                  textAlign: TextAlign.center,
+                  style: actionTextStyle()),
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_condition_combiner_bottomsheet_action_tile_builder.dart:45]:
```suggestion 
                child: Text(
                   actionName,
                   textAlign: TextAlign.center,
                   style: actionTextStyle()
                 ),
```

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+  final ConditionCombiner? conditionCombinerType;
+  final OnTapActionCallback? tapActionCallback;
+  final RuleFilterConditionScreenType ruleFilterConditionScreenType;
+
+  const RuleFilterTitle({
+    Key? key,
+    this.conditionCombinerType,
+    this.tapActionCallback,
+    required this.ruleFilterConditionScreenType,
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart:22]: `required ` parameters should be placed first (after `Key)`

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  @override
+  Widget build(BuildContext context) {
+    final conditionTitle = AppLocalizations.of(context).conditionTitleRulesFilter;
+
+    return Row(
+      crossAxisAlignment: CrossAxisAlignment.center,
+      mainAxisSize: MainAxisSize.min,
+      children: [
+        Text(
+          conditionTitle.substring(0, 2),
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart:34]: You should create separate text for it in `AppLocalization`

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            )
+            : DropDownButtonWidget<ConditionCombiner>(
+            items: ConditionCombiner.values,
+            itemSelected: conditionCombinerType,
+            supportSelectionIcon: true,
+            onChanged: (value) {
+              KeyboardUtils.hideKeyboard(context);
+              tapActionCallback?.call(value);
+            },
+          ),
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart:63]: indent

**Review by dab246 — COMMENTED** [review]
- Pay attention to code conventions
- Separate `styles/padding` into 1 style file to easily manage this.

**Comment by dab246** [comment]
Please test case send email to account config new rule filter. To check if the `condition/action` is working properly?

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart [hunk 5]
```diff
@@ -37,8 +37,8 @@ class RuleFilterConditionWidget extends StatelessWidget {
   @override
   Widget build(BuildContext context) {
     return Slidable(
-      // enabled: ruleFilterConditionScreenType == RuleFilterConditionScreenType.mobile ? true : false,
-      enabled: false,
+      enabled: ruleFilterConditionScreenType == RuleFilterConditionScreenType.mobile ? true : false,
+      // enabled: false,
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_condition_widget.dart:41]: remove it

File: lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+              )
+              : DropDownButtonWidget<ConditionCombiner>(
+              items: ConditionCombiner.values,
+              itemSelected: conditionCombinerType,
+              supportSelectionIcon: true,
+              onChanged: (value) {
+                KeyboardUtils.hideKeyboard(context);
+                tapActionCallback?.call(value);
+              },
+            ),
```
Comment (dab246) [inline lib/features/rules_filter_creator/presentation/widgets/rule_filter_title_builder.dart:59]: indent

File: rule_filter/lib/rule_filter/rule_condition_group.dart [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  RuleConditionGroup({
+    required this.conditionCombiner,
+    required this.conditions,
+  });
+
+  @override
+  List<Object?> get props => [
+        conditionCombiner,
+        conditions,
+      ];
```
Comment (dab246) [inline rule_filter/lib/rule_filter/rule_condition_group.dart:29]: indent

File: rule_filter/lib/rule_filter/tmail_rule.dart [hunk 8]
```diff
@@ -14,39 +15,54 @@ part 'tmail_rule.g.dart';
...
 
-  Map<String, dynamic> toJson() => _$TMailRuleToJson(this);
+  Map<String, dynamic> toJson() {
+    final result = _$TMailRuleToJson(this);
+
+    if (result['condition'] != null) {
+      result.remove('conditionGroup');
+    } else {
+      result.remove('condition');
+    }
```
Comment (dab246) [inline rule_filter/lib/rule_filter/tmail_rule.dart:40]: why remove it?

Comment (hieutbui) [reply rule_filter/lib/rule_filter/tmail_rule.dart:40]: cause condition and conditionGroup is null-safety and if one of this is null, the json data parse to api will include both. It is not valid.

Comment (dab246) [reply rule_filter/lib/rule_filter/tmail_rule.dart:40]: why not use `includeIfNull: false`? .

**Comment by hieutbui** [comment]
> Please test sending emails to accounts configured with the new rule filter. To check if the `condition/action` is working properly?

I updated the 2nd video for this case. Please watch it!!

**Review by dab246 — APPROVED** [review]
lgtm

**Review by hoangdat — APPROVED** [review]

**Comment by hoangdat** [comment]
Hi @Arsnael you can check it :))

**Pull request merged**